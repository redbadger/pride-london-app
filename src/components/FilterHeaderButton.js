// @flow
import React from "react";
import { StyleSheet, View } from "react-native";
import type { ViewStyleProp } from "react-native/Libraries/StyleSheet/StyleSheet";
import Text from "./Text";
import Touchable from "./Touchable";
import { filterButtonTextColor, whiteColor } from "../constants/colors";
import NumberBadge from "./NumberBadge";

type Props = {
  text: string,
  label: string,
  onPress: Function,
  onRef?: Function,
  style?: ViewStyleProp,
  badgeValue?: ?number,
  active: boolean
};

const FilterHeaderButton = ({
  text,
  onPress,
  onRef,
  style,
  badgeValue,
  active,
  label
}: Props) => (
  <Touchable
    style={[styles.button, style]}
    onPress={onPress}
    ref={onRef}
    accessibilityLabel={label}
  >
    <Text
      type="small"
      style={[
        styles.buttonText,
        active ? styles.activeColor : styles.inactiveColor
      ]}
    >
      {text}
    </Text>
    {badgeValue != null && (
      <View style={styles.badgeContainer}>
        <NumberBadge value={badgeValue} />
      </View>
    )}
  </Touchable>
);

FilterHeaderButton.defaultProps = {
  onRef: undefined,
  style: {},
  badgeValue: null
};

const styles = StyleSheet.create({
  button: {
    paddingHorizontal: 8,
    alignItems: "center",
    justifyContent: "center",
    flexDirection: "row"
  },
  buttonText: {
    fontFamily: "Roboto-Medium",
    textAlign: "center"
  },
  activeColor: {
    color: filterButtonTextColor
  },
  inactiveColor: {
    color: whiteColor
  },
  badgeContainer: {
    marginLeft: 6
  }
});

export default FilterHeaderButton;
