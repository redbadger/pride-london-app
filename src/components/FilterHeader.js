// @flow
import React from "react";
import { View, StyleSheet } from "react-native";
import DateFilterDialog from "./ConnectedDateFilterDialog";
import FilterHeaderButton from "./FilterHeaderButton";
import TimeFilterDialog from "./ConnectedTimeFilterDialog";
import Text from "./Text";
import {
  headerBgColor,
  interestButtonBgColor,
  interestButtonTextColor
} from "../constants/colors";
import text from "../constants/text";
import type { DateOrDateRange, Time } from "../data/date-time";
import { formatDateRange } from "../data/formatters";

type Props = {
  dateFilter: ?DateOrDateRange,
  timeFilter: Set<Time>
};

type State = {
  datesPickerVisible: boolean,
  timesPickerVisible: boolean
};

class FilterHeader extends React.PureComponent<Props, State> {
  state = {
    datesPickerVisible: false,
    timesPickerVisible: false
  };

  showDatePicker = () => {
    this.setState({ datesPickerVisible: true });
  };

  hideDatePicker = () => {
    this.setState({ datesPickerVisible: false });
  };

  showTimePicker = () => {
    this.setState({ timesPickerVisible: true });
  };

  hideTimePicker = () => {
    this.setState({ timesPickerVisible: false });
  };

  render() {
    const { dateFilter, timeFilter } = this.props;
    const formattedDateFilter = dateFilter
      ? formatDateRange(dateFilter)
      : text.anyDay;
    const timeArray = Array.from(timeFilter);
    const formattedTimeFilter =
      timeArray.length > 0 && timeArray.length < 3
        ? timeArray.map(time => text.time[time]).join(", ")
        : text.anyTime;

    return (
      <View style={styles.container}>
        <View testID="filter-header" style={styles.content}>
          <View style={styles.contentInterest}>
            <View style={styles.interestButton}>
              <Text type="h2" style={styles.interestButtonText}>
                {text.filterByInterest}
              </Text>
            </View>
            <View style={styles.mapButton}>
              <Text style={styles.mapButtonText}>Map</Text>
            </View>
          </View>
          <View style={styles.contentFilters}>
            <FilterHeaderButton
              text={formattedDateFilter}
              onPress={this.showDatePicker}
              style={styles.filterButton}
            />
            <FilterHeaderButton
              text={formattedTimeFilter}
              onPress={this.showTimePicker}
              style={styles.filterButton}
            />
            <FilterHeaderButton
              text={text.filters}
              onPress={() => {}}
              style={styles.filterButton}
            />
          </View>
        </View>
        <DateFilterDialog
          onApply={this.hideDatePicker}
          onCancel={this.hideDatePicker}
          visible={this.state.datesPickerVisible}
        />
        <TimeFilterDialog
          onApply={this.hideTimePicker}
          onCancel={this.hideTimePicker}
          visible={this.state.timesPickerVisible}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: headerBgColor
  },
  content: {
    paddingTop: 16,
    paddingBottom: 12
  },
  contentInterest: {
    alignItems: "center",
    flexDirection: "row"
  },
  interestButton: {
    flex: 1,
    height: 44,
    backgroundColor: interestButtonBgColor,
    marginLeft: 8,
    paddingHorizontal: 12,
    borderRadius: 4,
    justifyContent: "center"
  },
  interestButtonText: {
    color: interestButtonTextColor
  },
  mapButton: {
    marginHorizontal: 12,
    width: 52,
    height: 52,
    backgroundColor: interestButtonBgColor,
    alignItems: "center",
    justifyContent: "flex-end",
    borderRadius: 25
  },
  mapButtonText: {
    color: interestButtonTextColor,
    fontFamily: "Poppins-Bold",
    fontSize: 14,
    paddingBottom: 6
  },
  contentFilters: {
    flexDirection: "row",
    marginTop: 8
  },
  filterButton: {
    marginLeft: 8
  }
});

export default FilterHeader;
