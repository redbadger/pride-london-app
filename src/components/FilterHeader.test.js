// @flow
import React from "react";
import { shallow } from "enzyme";
import DateFilterDialog from "./ConnectedDateFilterDialog";
import FilterHeader from "./FilterHeader";
import FilterHeaderButton from "./FilterHeaderButton";
import type { Props as ComponentProps } from "./FilterHeader";

const render = (
  props: ComponentProps = {
    dateFilter: null,
    selectedCategories: new Set(),
    onFilterCategoriesPress: () => {},
    onFilterButtonPress: () => {},
    numTagFiltersSelected: 0
  }
) => shallow(<FilterHeader {...props} />);

describe("renders correctly", () => {
  it("with any date and any time", () => {
    const output = render({
      dateFilter: null,
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      numTagFiltersSelected: 0
    });
    expect(output).toMatchSnapshot();
  });

  it("with any date and any time (empty time set)", () => {
    const output = render({
      dateFilter: null,
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      numTagFiltersSelected: 0
    });
    expect(output).toMatchSnapshot();
  });

  it("with single date and single time", () => {
    const output = render({
      dateFilter: { startDate: "2018-02-02", endDate: "2018-02-02" },
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      numTagFiltersSelected: 0
    });
    expect(output).toMatchSnapshot();
  });

  it("with date range and two times", () => {
    const output = render({
      dateFilter: {
        startDate: "2018-02-02",
        endDate: "2018-02-03"
      },
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      numTagFiltersSelected: 0
    });
    expect(output).toMatchSnapshot();
  });

  it("with tag filters selected", () => {
    const output = render({
      dateFilter: null,
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      numTagFiltersSelected: 2
    });
    expect(output).toMatchSnapshot();
  });
});

it("opens date picker when users presses date filter button", () => {
  const output = render();
  const filterDateButton = output.find(FilterHeaderButton).at(0);
  filterDateButton.simulate("press");

  expect(output.state("datesPickerVisible")).toBe(true);
});

it("closes date picker when users presses applies changes", () => {
  const output = render();
  output.setState({ datesPickerVisible: true });

  const datePicker = output.find(DateFilterDialog);
  datePicker.simulate("apply");

  expect(output.state("datesPickerVisible")).toBe(false);
});
