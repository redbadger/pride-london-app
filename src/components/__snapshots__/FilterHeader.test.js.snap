// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with any date and any time (empty time set) 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#A5A5A5",
    }
  }
>
  <View
    style={
      Object {
        "paddingBottom": 12,
        "paddingTop": 16,
      }
    }
    testID="filter-header"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#545454",
            "borderRadius": 4,
            "flex": 1,
            "height": 44,
            "justifyContent": "center",
            "marginLeft": 8,
            "paddingHorizontal": 12,
          }
        }
      >
        <Text
          markdown={false}
          style={
            Object {
              "color": "#FFFFFF",
            }
          }
          type="h2"
        >
          I'm interested in...
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#545454",
            "borderRadius": 25,
            "height": 52,
            "justifyContent": "flex-end",
            "marginHorizontal": 12,
            "width": 52,
          }
        }
      >
        <Text
          markdown={false}
          style={
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Poppins-Bold",
              "fontSize": 14,
              "paddingBottom": 6,
            }
          }
          type="text"
        >
          Map
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 8,
        }
      }
    >
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="Any day"
      />
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="Any time"
      />
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="Filters"
      />
    </View>
  </View>
  <Connect(DateRangePickerDialog)
    onApply={[Function]}
    onCancel={[Function]}
    visible={false}
  />
  <Connect(MultiSelectDialog)
    onApply={[Function]}
    onCancel={[Function]}
    visible={false}
  />
</View>
`;

exports[`renders correctly with any date and any time 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#A5A5A5",
    }
  }
>
  <View
    style={
      Object {
        "paddingBottom": 12,
        "paddingTop": 16,
      }
    }
    testID="filter-header"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#545454",
            "borderRadius": 4,
            "flex": 1,
            "height": 44,
            "justifyContent": "center",
            "marginLeft": 8,
            "paddingHorizontal": 12,
          }
        }
      >
        <Text
          markdown={false}
          style={
            Object {
              "color": "#FFFFFF",
            }
          }
          type="h2"
        >
          I'm interested in...
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#545454",
            "borderRadius": 25,
            "height": 52,
            "justifyContent": "flex-end",
            "marginHorizontal": 12,
            "width": 52,
          }
        }
      >
        <Text
          markdown={false}
          style={
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Poppins-Bold",
              "fontSize": 14,
              "paddingBottom": 6,
            }
          }
          type="text"
        >
          Map
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 8,
        }
      }
    >
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="Any day"
      />
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="Any time"
      />
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="Filters"
      />
    </View>
  </View>
  <Connect(DateRangePickerDialog)
    onApply={[Function]}
    onCancel={[Function]}
    visible={false}
  />
  <Connect(MultiSelectDialog)
    onApply={[Function]}
    onCancel={[Function]}
    visible={false}
  />
</View>
`;

exports[`renders correctly with date range and two times 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#A5A5A5",
    }
  }
>
  <View
    style={
      Object {
        "paddingBottom": 12,
        "paddingTop": 16,
      }
    }
    testID="filter-header"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#545454",
            "borderRadius": 4,
            "flex": 1,
            "height": 44,
            "justifyContent": "center",
            "marginLeft": 8,
            "paddingHorizontal": 12,
          }
        }
      >
        <Text
          markdown={false}
          style={
            Object {
              "color": "#FFFFFF",
            }
          }
          type="h2"
        >
          I'm interested in...
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#545454",
            "borderRadius": 25,
            "height": 52,
            "justifyContent": "flex-end",
            "marginHorizontal": 12,
            "width": 52,
          }
        }
      >
        <Text
          markdown={false}
          style={
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Poppins-Bold",
              "fontSize": 14,
              "paddingBottom": 6,
            }
          }
          type="text"
        >
          Map
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 8,
        }
      }
    >
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="2 Feb - 3 Feb"
      />
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="Morning, Afternoon"
      />
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="Filters"
      />
    </View>
  </View>
  <Connect(DateRangePickerDialog)
    onApply={[Function]}
    onCancel={[Function]}
    visible={false}
  />
  <Connect(MultiSelectDialog)
    onApply={[Function]}
    onCancel={[Function]}
    visible={false}
  />
</View>
`;

exports[`renders correctly with single date and single time 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#A5A5A5",
    }
  }
>
  <View
    style={
      Object {
        "paddingBottom": 12,
        "paddingTop": 16,
      }
    }
    testID="filter-header"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#545454",
            "borderRadius": 4,
            "flex": 1,
            "height": 44,
            "justifyContent": "center",
            "marginLeft": 8,
            "paddingHorizontal": 12,
          }
        }
      >
        <Text
          markdown={false}
          style={
            Object {
              "color": "#FFFFFF",
            }
          }
          type="h2"
        >
          I'm interested in...
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#545454",
            "borderRadius": 25,
            "height": 52,
            "justifyContent": "flex-end",
            "marginHorizontal": 12,
            "width": 52,
          }
        }
      >
        <Text
          markdown={false}
          style={
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Poppins-Bold",
              "fontSize": 14,
              "paddingBottom": 6,
            }
          }
          type="text"
        >
          Map
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 8,
        }
      }
    >
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="2 Feb"
      />
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="Afternoon"
      />
      <FilterHeaderButton
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 8,
          }
        }
        text="Filters"
      />
    </View>
  </View>
  <Connect(DateRangePickerDialog)
    onApply={[Function]}
    onCancel={[Function]}
    visible={false}
  />
  <Connect(MultiSelectDialog)
    onApply={[Function]}
    onCancel={[Function]}
    visible={false}
  />
</View>
`;
